File: apps/linux/src/app/playback/languages.rs
//! Display common ISO 639-1/2 codes without discarding unknown language tags.
pub fn name(code: &str) -> String {
    let normalized = code.trim().replace('_', "-").to_ascii_lowercase();
    let mut tags = normalized.split('-');
    let language = tags.next().unwrap_or_default();
    let name = match language {
        "en" | "eng" => "English",
        "fr" | "fra" | "fre" => "French",
        "de" | "deu" | "ger" => "German",
        "es" | "spa" => "Spanish",
        "pt" | "por" => "Portuguese",
        "it" | "ita" => "Italian",
        "ru" | "rus" => "Russian",
        "uk" | "ukr" => "Ukrainian",
        "ja" | "jpn" => "Japanese",
        "ko" | "kor" => "Korean",
        "zh" | "zho" | "chi" => "Chinese",
        "ar" | "ara" => "Arabic",
        "hi" | "hin" => "Hindi",
        "mr" | "mar" => "Marathi",
        "ta" | "tam" => "Tamil",
        "te" | "tel" => "Telugu",
        "ml" | "mal" => "Malayalam",
        "kn" | "kan" => "Kannada",
        "bn" | "ben" => "Bengali",
        "gu" | "guj" => "Gujarati",
        "pa" | "pan" => "Punjabi",
        "ur" | "urd" => "Urdu",
        "or" | "ori" => "Odia",
        "as" | "asm" => "Assamese",
        "ne" | "nep" => "Nepali",
        "si" | "sin" => "Sinhala",
        "th" | "tha" => "Thai",
        "vi" | "vie" => "Vietnamese",
        "id" | "ind" => "Indonesian",
        "ms" | "msa" | "may" => "Malay",
        "tl" | "tgl" | "fil" => "Filipino",
        "tr" | "tur" => "Turkish",
        "fa" | "fas" | "per" => "Persian",
        "he" | "heb" | "iw" => "Hebrew",
        "pl" | "pol" => "Polish",
        "nl" | "nld" | "dut" => "Dutch",
        "sv" | "swe" => "Swedish",
        "no" | "nor" => "Norwegian",
        "nb" | "nob" => "Norwegian Bokmål",
        "nn" | "nno" => "Norwegian Nynorsk",
        "da" | "dan" => "Danish",
        "fi" | "fin" => "Finnish",
        "el" | "ell" | "gre" => "Greek",
        "cs" | "ces" | "cze" => "Czech",
        "sk" | "slk" | "slo" => "Slovak",
        "hu" | "hun" => "Hungarian",
        "ro" | "ron" | "rum" => "Romanian",
        "bg" | "bul" => "Bulgarian",
        "hr" | "hrv" => "Croatian",
        "sr" | "srp" => "Serbian",
        "sl" | "slv" => "Slovenian",
        "bs" | "bos" => "Bosnian",
        "sq" | "sqi" | "alb" => "Albanian",
        "mk" | "mkd" | "mac" => "Macedonian",
        "et" | "est" => "Estonian",
        "lv" | "lav" => "Latvian",
        "lt" | "lit" => "Lithuanian",
        "is" | "isl" | "ice" => "Icelandic",
        "ca" | "cat" => "Catalan",
        "eu" | "eus" | "baq" => "Basque",
        "gl" | "glg" => "Galician",
        "ka" | "kat" | "geo" => "Georgian",
        "hy" | "hye" | "arm" => "Armenian",
        "az" | "aze" => "Azerbaijani",
        "kk" | "kaz" => "Kazakh",
        "uz" | "uzb" => "Uzbek",
        "sw" | "swa" => "Swahili",
        "af" | "afr" => "Afrikaans",
        "zu" | "zul" => "Zulu",
        "und" => "Unspecified language",
        "mul" => "Multiple languages",
        "zxx" => "No spoken language",
        "" => return String::new(),
        _ => return code.trim().to_string(),
    };
    let qualifiers: Vec<_> = tags
        .map(|tag| match tag {
            "us" => "United States".into(),
            "gb" => "United Kingdom".into(),
            "br" => "Brazil".into(),
            "pt" => "Portugal".into(),
            "ca" => "Canada".into(),
            "mx" => "Mexico".into(),
            "hans" => "Simplified".into(),
            "hant" => "Traditional".into(),
            "latn" => "Latin".into(),
            "cyrl" => "Cyrillic".into(),
            other => other.to_ascii_uppercase(),
        })
        .collect();
    if qualifiers.is_empty() {
        name.into()
    } else {
        format!("{name} ({})", qualifiers.join(", "))
    }
}
#[cfg(test)]
mod tests {
    #[test]
    fn displays_codes_and_preserves_unknowns() {
        for code in ["en", "EN", "eng"] {
            assert_eq!(super::name(code), "English");
        }
        assert_eq!(super::name("pt-BR"), "Portuguese (Brazil)");
        assert_eq!(super::name("hi"), "Hindi");
        assert_eq!(super::name("unknown-tag"), "unknown-tag");
    }
}

File: crates/madari-tv/web/src/languages.ts
/**
 * Mirrors the Rust `madari_native::playback::languages` labels so the browser and
 * the TV show the same names for a language code.
 */
const NAMES: Record<string, string> = {
  en: 'English', eng: 'English', hi: 'Hindi', hin: 'Hindi', ta: 'Tamil', tam: 'Tamil',
  te: 'Telugu', tel: 'Telugu', ml: 'Malayalam', mal: 'Malayalam', kn: 'Kannada', kan: 'Kannada',
  mr: 'Marathi', mar: 'Marathi', bn: 'Bengali', ben: 'Bengali', gu: 'Gujarati', guj: 'Gujarati',
  pa: 'Punjabi', pan: 'Punjabi', ur: 'Urdu', urd: 'Urdu', as: 'Assamese', asm: 'Assamese',
  or: 'Odia', ori: 'Odia', ne: 'Nepali', nep: 'Nepali', si: 'Sinhala', sin: 'Sinhala',
  fr: 'French', fra: 'French', fre: 'French', de: 'German', deu: 'German', ger: 'German',
  es: 'Spanish', spa: 'Spanish', pt: 'Portuguese', por: 'Portuguese', it: 'Italian', ita: 'Italian',
  nl: 'Dutch', nld: 'Dutch', dut: 'Dutch', pl: 'Polish', pol: 'Polish', ru: 'Russian', rus: 'Russian',
  uk: 'Ukrainian', ukr: 'Ukrainian', tr: 'Turkish', tur: 'Turkish', sv: 'Swedish', swe: 'Swedish',
  da: 'Danish', dan: 'Danish', fi: 'Finnish', fin: 'Finnish', no: 'Norwegian', nor: 'Norwegian',
  nb: 'Norwegian Bokmål', nn: 'Norwegian Nynorsk', el: 'Greek', ell: 'Greek', gre: 'Greek',
  cs: 'Czech', ces: 'Czech', sk: 'Slovak', slk: 'Slovak', hu: 'Hungarian', hun: 'Hungarian',
  ro: 'Romanian', ron: 'Romanian', bg: 'Bulgarian', bul: 'Bulgarian', hr: 'Croatian', hrv: 'Croatian',
  sr: 'Serbian', srp: 'Serbian', sl: 'Slovenian', slv: 'Slovenian', bs: 'Bosnian', bos: 'Bosnian',
  sq: 'Albanian', mk: 'Macedonian', et: 'Estonian', lv: 'Latvian', lt: 'Lithuanian',
  is: 'Icelandic', ca: 'Catalan', eu: 'Basque', gl: 'Galician', ka: 'Georgian', hy: 'Armenian',
  az: 'Azerbaijani', kk: 'Kazakh', uz: 'Uzbek', sw: 'Swahili', af: 'Afrikaans', zu: 'Zulu',
  ja: 'Japanese', jpn: 'Japanese', ko: 'Korean', kor: 'Korean', zh: 'Chinese', zho: 'Chinese',
  chi: 'Chinese', ar: 'Arabic', ara: 'Arabic', he: 'Hebrew', heb: 'Hebrew', fa: 'Persian',
  fas: 'Persian', per: 'Persian', th: 'Thai', tha: 'Thai', vi: 'Vietnamese', vie: 'Vietnamese',
  id: 'Indonesian', ind: 'Indonesian', ms: 'Malay', msa: 'Malay', may: 'Malay',
  tl: 'Filipino', tgl: 'Filipino', fil: 'Filipino',
  und: 'Unspecified language', mul: 'Multiple languages', zxx: 'No spoken language',
};

const QUALIFIERS: Record<string, string> = {
  us: 'United States', gb: 'United Kingdom', br: 'Brazil', pt: 'Portugal', ca: 'Canada', mx: 'Mexico',
  hans: 'Simplified', hant: 'Traditional', latn: 'Latin', cyrl: 'Cyrillic',
};

export function languageName(code: string): string {
  const normalized = code.trim().replace(/_/g, '-').toLowerCase();
  if (!normalized) return '';
  const [language, ...qualifiers] = normalized.split('-');
  const base = NAMES[language] ?? code.trim();
  if (!qualifiers.length) return base;
  const suffix = qualifiers.map((tag) => QUALIFIERS[tag] ?? tag.toUpperCase()).join(', ');
  return `${base} (${suffix})`;
}

/** Curated list offered when adding a language; unknown tags still round-trip. */
export const COMMON_LANGUAGES = [
  'en', 'hi', 'ta', 'te', 'ml', 'kn', 'mr', 'bn', 'gu', 'pa', 'ur', 'as', 'or', 'ne', 'si',
  'fr', 'de', 'es', 'pt', 'it', 'nl', 'pl', 'ru', 'uk', 'tr', 'sv', 'da', 'fi', 'no', 'el',
  'ja', 'ko', 'zh', 'ar', 'he', 'fa', 'th', 'vi', 'id', 'ms', 'tl',
];

export const TRACK_PREFERENCES: Array<{key: 'subtitle_sdh' | 'subtitle_forced' | 'audio_description' | 'audio_commentary'; label: string}> = [
  {key: 'subtitle_sdh', label: 'SDH subtitles'},
  {key: 'subtitle_forced', label: 'Forced subtitles'},
  {key: 'audio_description', label: 'Audio description'},
  {key: 'audio_commentary', label: 'Commentary tracks'},
];
